### PR TITLE
[TS migration] Migrate SearchPage

### PR DESCRIPTION
**Pull request by ruben-rebelo**
### Details
[TS migration] Migrate _**SearchPage**_ to TypeScript

### Fixed Issues
$ https://github.com/Expensify/App/issues/25182
PROPOSAL: N/A

### Tests
- [x] Verify that no errors appear in the JS console

**Test NewChatPage**:

- Log into an account
- On the initial screens tap on search button on the drawer menu
- This screen should work as before together with search, footer and select user functionality

### Offline tests
N/A

### QA Steps
- [x] Verify that no errors appear in the JS console

Same as in the **Tests** section.

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

android-native

</details>

<details>
<summary>Android: mWeb Chrome</summary>

android-web

</details>

<details>
<summary>iOS: Native</summary>

ios-native

</details>

<details>
<summary>iOS: mWeb Safari</summary>

ios-web

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

macos-web

</details>

<details>
<summary>MacOS: Desktop</summary>

macos-native

</details>

### CODE REVIEW

File: src/pages/SearchPage.tsx
Comment (VickyStash) [inline src/pages/SearchPage.tsx:22]:
```suggestion
type SearchPageOnyxProps = {
```

File: src/pages/SearchPage.tsx
Comment (VickyStash) [inline src/pages/SearchPage.tsx:34]: This way it will be easier add props if needed in the future and the structure will be the same as in other places
```suggestion
};

type SearchPageProps = SearchPageOnyxProps;
```

File: src/pages/SearchPage.tsx
Comment (VickyStash) [inline src/pages/SearchPage.tsx:201]:
```suggestion
export default withOnyx<SearchPageProps, SearchPageOnyxProps>({
```

File: src/pages/SearchPage.tsx
Comment (VickyStash) [inline src/pages/SearchPage.tsx:36]: I guess this type should have similar typing with [this one](https://github.com/Expensify/App/pull/32470/files#diff-7d43f3c3e4decdb0057928e895f456d51c0f4081837edd5e2c4ed7ddd570e274R118), cause it uses props from there

File: src/pages/SearchPage.tsx
Comment (VickyStash) [inline src/pages/SearchPage.tsx:39]: Just an empty array? it doesn't look right for the type, also are you sure it's array at all?

File: src/pages/SearchPage.tsx
Comment (VickyStash) [inline src/pages/SearchPage.tsx:47]: Are you sure `userToInvite` is an array?

File: src/pages/SearchPage.tsx
Comment (VickyStash) [inline src/pages/SearchPage.tsx:152]: Maybe this way it will be more safe? What do you think?
```suggestion
            Report.navigateToAndOpenReport(option.login ? [option.login] : []);
```

File: src/pages/SearchPage.tsx
Comment (VickyStash) [reply src/pages/SearchPage.tsx:34]: I've just realized it's a page, so it should have navigation props, so it can look like this
```suggestion
};

type SearchPageProps = SearchPageOnyxProps & StackScreenProps<SearchNavigatorParamList, typeof SCREENS.SEARCH_ROOT>;
```

File: src/pages/SearchPage.tsx
Comment (VickyStash) [inline src/pages/SearchPage.tsx:33]:
```suggestion
    /** All reports shared with the user */
    reports: OnyxCollection<OnyxTypes.Report>;
```

File: src/pages/SearchPage.tsx
Comment (fabioh8010) [reply src/pages/SearchPage.tsx:36]: @ruben-rebelo how about this comment?

File: src/pages/SearchPage.tsx
Comment (fabioh8010) [inline src/pages/SearchPage.tsx:100]: Add return type for this function

File: src/pages/SearchPage.tsx
Comment (VickyStash) [inline src/pages/SearchPage.tsx:73]:
```suggestion
        } = OptionsListUtils.getSearchOptions(reports, personalDetails, searchValue.trim(), betas ?? []);
```

**Review by blazejkustra — APPROVED** [review]

**Comment by ruben-rebelo** [comment]
Hello @ntdiary, no problem!
I've resolved the conflicts, please have a look!

**Review by ntdiary — COMMENTED** [review]
Hi, the PR looks good. However, I found we have another PR #35058 in progress to remove this file and use the new `SearchPage/index.js`. Also, considering we have a TODO comment waiting for the `OptionsSelector` migration, I'm not sure how we should proceed here. Maybe we could wait for PR #35058 to be completed and then migrate `SearchPage/index.js`? 
I've posted a message on slack to get some inputs, and please feel free to let me know If you have a different thought ❤️

**Comment by ruben-rebelo** [comment]
Hello @ntdiary, I didn't knew about that PR, Let's keep this on hold until that one is merged than I proceed with the migration from there. Thank you for the information!

**Comment by ruben-rebelo** [comment]
Hello @ntdiary, I've added the changes after the refactor has been done.
I've tested again to ensure everything was working as expected.
Please take a look.

**Comment by ruben-rebelo** [comment]
Hello @ntdiary, I've added the changes after the refactor has been done.
I've tested again to ensure everything was working as expected.
Please take a look.

**Comment by ruben-rebelo** [comment]
Hello @ntdiary, I've migrated the SearchPage after the refactor.
I've tested the page to ensure it was working as before.
Please have a look, thank you!

File: src/components/SelectionList/types.ts
Comment (fabioh8010) [reply src/components/SelectionList/types.ts:50]: Well, if `null` is really needed in `ReportUtils.OptionData` we should support, otherwise let's delete it so we don't need to cascade this `?: boolean | null;` up to this component! cc @ruben-rebelo

Comment (ruben-rebelo) [reply src/components/SelectionList/types.ts]: I agree, I've reverted all the changes made on SelectionList and updated SearchPage.
Please have a look @ntdiary

Comment (ruben-rebelo) [reply src/components/SelectionList/types.ts]: I'm afraid with this changed we can't avoid the assertion usage.
Do you prefer to revert this one and keep the nulls for this PR?
I'll update the offlineMessage and textInputHint with these suggestions.

**Review by ntdiary — CHANGES_REQUESTED** [review]
@ruben-rebelo, there's a ts error in `WorkspaceMembersPage`, and I've also made a few minor ts suggestions. I have tested it locally and they works well. I will fill out the list after they are addressed. :)

**Comment by ruben-rebelo** [comment]
@ntdiary I've updated this PR including your suggestions + merge with main + some ts fixes that appeared after the changes.
Please have a look

**Review by ntdiary — CHANGES_REQUESTED** [review]
@ruben-rebelo, the last minor change, I'm testing it now, and will update the videos later. 😄

**Review by ntdiary — APPROVED** [review]
LGTM :)

**Review by yuwenmemon — DISMISSED** [review]

**Comment by yuwenmemon** [comment]
@ruben-rebelo Conflicts

**Comment by ruben-rebelo** [comment]
@yuwenmemon Conflicts Resolved

**Review by yuwenmemon — APPROVED** [review]

**Comment by yuwenmemon** [comment]
Apologies for the delay! I was sick last week 🤕

**Pull request merged**
